core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveRedundantDistinctAggregation.java

```diff
@@ -0,0 +1,105 @@
+
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.iterative.Lookup;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.AggregationNode;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.ProjectNode;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static io.trino.sql.planner.plan.Patterns.aggregation;
+
+/**
+ * Removes DISTINCT only aggregation, when the input source is already distinct over a subset of
+ * the grouping keys as a result of another aggregation.
+ *
+ * Given:
+ * <pre>
+ * - Aggregate[keys = [a, max]]
+ *   - Aggregate[keys = [a]]
+ *     max := max(b)
+ * </pre>
+ * <p>
+ * Produces:
+ * <pre>
+ *   - Aggregate[keys = [a]]
+ *     max := max(b)
+ * </pre>
+ */
+public class RemoveRedundantDistinctAggregation
+        implements Rule<AggregationNode>
+{
+    private static final Pattern<AggregationNode> PATTERN = aggregation()
+            .matching(RemoveRedundantDistinctAggregation::isDistinctOnlyAggregation);
+
+    @Override
+    public Pattern<AggregationNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(AggregationNode aggregationNode, Captures captures, Context context)
+    {
+        Lookup lookup = context.getLookup();
+        if (isDistinctOverGroupingKeys(lookup.resolve(aggregationNode.getSource()), lookup, new HashSet<>(aggregationNode.getGroupingKeys())) != null) {
+            return Result.ofPlanNode(aggregationNode.getSource());
+        }
+        else {
+            return Result.empty();
+        }
+    }
+
+    private static boolean isDistinctOnlyAggregation(AggregationNode node)
+    {
+        return node.producesDistinctRows() && node.getGroupingSetCount() == 1;
+    }
+
+    private static List<Symbol> isDistinctOverGroupingKeys(PlanNode node, Lookup lookup, Set<Symbol> groupingKeys)
+    {
+        return switch (node) {
+            case AggregationNode aggregationNode -> {
+                if (aggregationNode.getGroupingSets().getGroupingSetCount() == 1 && groupingKeys.containsAll(aggregationNode.getGroupingSets().getGroupingKeys())) {
+                    yield aggregationNode.getGroupingSets().getGroupingKeys();
+                }
+                yield null;
+            }
+
+            // Project nodes introduce new symbols for computed expressions, and therefore end up preserving distinctness
+            // between the distinct aggregation and the child aggregation nodes so long as all child aggregation keys
+            // remain present (without transformation by the project) in the distinct aggregation grouping keys
+            case ProjectNode projectNode -> {
+                List<Symbol> childGroupingKeys = isDistinctOverGroupingKeys(lookup.resolve(projectNode.getSource()), lookup, groupingKeys);
```

Comment:
This is still not quite correct. It's technically not legal to compare the grouping keys from the leaf aggregation with the grouping keys passed from above without performing the reverse translations defined by this projection, as the meaning of symbols is local to each pair of operations and it may have been changed by the projection.

Reply:
It may not _locally_ be correct to perform the check: `groupingKeys.containsAll(aggregationNode.getGroupingSets().getGroupingKeys())` in isolation, since we don't know at that point whether the grouping key symbols from the parent and child aggregations have the same meaning. However, when we return the leaf aggregation symbols back up the call stack, we verfiy that any `ProjectNode`s encountered only perform identity projections for all leaf aggregation symbols- at which point we can say that they _do_ have the same meaning by the time we reach the parent aggregation again and the transform should be considered to be valid.

Reply:
My concern is that it's too brittle and it relies on non-local assumptions. The call to `isDistinctOverGroupingKeys(lookup.resolve(projectNode.getSource()), lookup, groupingKeys)` is incorrect by construction, and whether `childGroupingKeys.stream().allMatch(projectNode.getAssignments()::isIdentity)` is sufficient to compensate depends on knowledge that this specific callsite doesn't have. For example, adding support of another type of node might break these assumptions. 

Also, in the call to `childGroupingKeys.stream().allMatch(projectNode.getAssignments()::isIdentity)`, it's incorrect to perform a lookup in the project node assignments using the symbols from the input scope as the key. The keys belong in the "output" scope. Today that works, but if in the future we make such lookups stricter by failing if the key is not from the proper scope, this will break.

Separately, the check is too specific. For example, it won't be able to catch plans like these ones:

```
- Aggregate [k]
  - Project k = x
    - Aggregate [x]
```

Reply:
> My concern is that it's too brittle and it relies on non-local assumptions

That’s fair, although I would point out that the assumptions are self contained to the single static private method, so they’re reasonably localized.

> For example, adding support of another type of node might break these assumptions.

Agreed. Maybe we can add a comment to clarify the specific assumptions being made and warn against expanding the rule to accept new intervening node types?

> The keys belong in the "output" scope. Today that works, but if in the future we make such lookups stricter by failing if the key is not from the proper scope, this will break.

It’s hard for me to guess what that might look like, but keys of interest by definition straddle the input and output scope, and the check for identity transforms guarantees that they are the same. (This is related to the next point, more on that there).

> Separately, the check is too specific. For example, it won't be able to catch plans like these ones:

> ```
- Aggregate [k]
  - Project k = x
    - Aggregate [x]
```

It won’t detect that, true- but it’s a missed optimization opportunity at that point and not a correctness bug. I’m not sure whether there are any plan optimizers today that would create new symbols for an identity assignment, but if there are then I guess we should plan for that? It’s definitely going to be more expensive to find any identity mappings from all input symbols and consider the intersection of those with the parent grouping symbols, but should be doable at that higher cost. Is that the recommendation here?

Reply:
Ok, fine. Let's add that documentation and get it in as is. We can fix those issue later when they come up. Also, please add a comment explaining that the lookups are not "proper" and that we're taking a shortcut.

Reply:
Thank you for all the comments. I have rewritten the method to include translating the grouping set symbols for project nodes.